Abort compilations for exceptions during invokedynamic resolution

nbhuiyan opened the pull request:
Invokedynamic call resolution can either result in a valid entry
in the CallSite table if successful, or an exception object
otherwise. The call site table entry is a two element array
containing the MemberName and appendix objects necessary for
constructing a resolved invokedynamic call. The exception object
would not be an array type object, and therefore it is sufficient
to check if the entry obtained from the CallSite table is an array
instance to determine if the resolution was successful.

This changeset handles the cases when we have an invalid entry
in the CallSite table that results in either failed inlining or
failed compilation, depending on whether the invokedynamic is part
of the inlining candidate or the root method, respectively.

Issue: #16965

nbhuiyan commented:
I believe the new VM API being added will need a JITServer override. I will try to implement that as well.

0xdaryl commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -279,6 +279,16 @@ class TR_J9VMBase : public TR_FrontEnd
+    * \return true if object is an array class instance
+    * \return false if object is not an array class instance
+    */
+   virtual bool objectIsArray(J9VMThread *currentThread, TR_OpaqueClassBlock *object);
```
You may have chosen this name because it matches the helper you end up calling, but since this is an inquiry API in the FrontEnd would it be better to call it `isArrayObject()` instead?

nbhuiyan commented:
@0xdaryl 
>  Is this issue only with the JIT's (mis)understanding of invokedynamic and the fix applies to both the OpenJ9 MH implementation as well as the OpenJDK LambdaForm MH implementations?

No, this fix only applies to the OpenJDK MH implementation. The older OpenJ9 MH implementation has a different mechanism for invokedynamic resolution, and does not place an exception object in the CallSite table entry, and the valid table entries themselves are not arrays.

nbhuiyan commented:
@0xdaryl I have addressed your initial review comments. There are some JITServer test failures due to the missing JITServer implementation. I am currently working on that.

0xdaryl commented:
> No, this fix only applies to the OpenJDK MH implementation. The older OpenJ9 MH implementation has a different mechanism for invokedynamic resolution, and does not place an exception object in the CallSite table entry, and the valid table entries themselves are not arrays.

OK, I just expanded the code around the fix and saw the #ifdefs guarding it, which is what I was going to ask you to add.  So it is all good.

Tag me when you think this PR is ready for another review.

nbhuiyan commented:
The JITServer test failures I observed before in my personal build now appear to be passing. @0xdaryl Ready for review.

@mpirvu I'd appreciate your review of the JITServer changes

0xdaryl commented:
Jenkins test sanity,sanity.openjdk all jdk21

mpirvu reviewed (commented):
LGTM
Since a new message has been added, could you please increment the MINOR_NUMBER here
https://github.com/eclipse-openj9/openj9/blob/master/runtime/compiler/net/CommunicationStream.hpp#L121

Also, I was wondering why the parameter for `isArrayObject(TR_OpaqueClassBlock *object);` is declared to be as a pointer to a class, when in fact it's an object.

nbhuiyan commented:
@mpirvu I have updated the MINOR_NUMBER

> I was wondering why the parameter for isArrayObject(TR_OpaqueClassBlock *object); is declared to be as a pointer to a class, when in fact it's an object

Thanks for pointing that out, I should not be using TR_OpaqueClassBlock* type for an object. Fixed that as well.

mpirvu commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6605,6 +6605,14 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
+      // and appendix objects necessary for constructing a resolved invokedynamic adapter method call.
+      // The exception object would not be an array type object, and therefore it is sufficient to check if
+      // the entry obtained from the CallSite table is an array instance to determine if the resolution was successful.
+      if (!fej9()->isArrayObject(*invokeCacheArray))
```
I am having second thoughts about this one. if `*invokeCacheArray` is truly a pointer to an object, what prevents GC from moving that object while we try to access it? We may need to hold VMAccess while executing `if (!fej9()->isArrayObject(*invokeCacheArray))` (extend the `VMAccessCriticalSection` which starts below).
There is also a comment below that "// this will not work in AOT or JITServer", so probably there is some guarantee that we cannot enter this path for JITServer.

nbhuiyan replied:
> if *invokeCacheArray is truly a pointer to an object, what prevents GC from moving that object while we try to access it?

I did not feel it was necessary to ensure VM access for this. My reasoning was, immediately after checking if the corresponding entry in the call site table had a valid object pointer using

https://github.com/eclipse-openj9/openj9/blob/ab5dfcfca290227c639a8d6877ca11773e723a42/runtime/compiler/env/j9method.cpp#L5915-L5919

We want to check whether that non-null object pointer cached in the call site table slot was an array or not. If that object pointer at `ramClass->callSites[callSiteIndex]` (i.e, `*invokeCacheArray`) is not pointing to a valid array object, then we should abort anyway. The slot is populated by an object that the VM's invokedynamic resolution mechanism allocates and (I'm speculating) the GC probably can not move. It's not a regular Java object. Once a call site table slot is populated (only if it was NULL to begin with), it is not modified again. What we do need VM access for however is accessing the individual elements in the `invokeCacheArray`, which is why we have the VM access being acquired right after this.

Based on my reasoning, do you still think it is necessary to put the isArrayObject check within the block with VM access acquired?

> There is also a comment below that "// this will not work in AOT or JITServer", so probably there is some guarantee that we cannot enter this path for JITServer.

This was true when this code was first written during the initial OpenJDK MethodHandle implementation, but since then JITServer became capable of supporting OpenJDK MHs and I can confirm that this code path is executed in JITServer now based on my testing. I will update the comment here to reflect the current state.

nbhuiyan replied:
Responding to your overall review comment:

> Moreover, the JITServer should never receive any object references. Thus, the new frontend method should just fatal assert if executed at the server rather than sending a message at the client.

Wouldn't this mean that JITServer loses support for any method containing `invokedynamic` bytecodes?

mpirvu replied:
If that array stored in one of the callsites is really special and cannot be moved by GC (maybe because it does not reside in the heap), then you don't need VM access for checking whether it's an array. Interesting though that the two elements of the array are protected by VM access.

mpirvu replied:
`fej9()->targetMethodFromMemberName()` has a FATAL_ASSERT in the JITServer implementation.
```
TR_OpaqueMethodBlock*
TR_J9ServerVM::targetMethodFromMemberName(uintptr_t memberName)
   {
   TR_ASSERT_FATAL(false, "targetMethodFromMemberName must not be called on JITServer");
   return NULL;
   }
   ```
   so if we ever entered that path while compiling at the server, the assert would be triggered.

mpirvu replied:
Another point: `getResolvedDynamicMethod()` has a different implementation for JITServer where the server asks the client to execute the corresponding method and then, with the answers received from the client, the server builds a resolvedMethod. Thus, the server does not need to access Java objects.

nbhuiyan replied:
Thank you for pointing those out, I had the wrong understanding of how things were being handled in JITServer mode. I will make the changes as you suggested.

I will also enforce VM access requirement for checking if object is array, so as not to cause confusion in the future for others, as there is no real benefit to doing this without VM access but lots of potential danger if this API was later used to check if heap objects were array types.

mpirvu commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -1256,6 +1256,12 @@ InterpreterEmulator::visitInvokedynamic()
+   // CallSite table entry is expected to be an array object upon successful resolution, but this is not
+   // the case when an exception occurs during the invokedynamic resolution, in which case an exception
+   // object is placed in the slot instead
+   if (!comp()->fej9()->isArrayObject(*invokeCacheArray))
```
Here too will need to have VM access while looking inside the object `*invokeCacheArray`. 
We acquire VM access later on in 
`fej9->getKnotIndexOfInvokeCacheArrayAppendixElement(comp(), invokeCacheArray);`
called from
`updateKnotAndCreateCallSiteUsingInvokeCacheArray`
Hence, maybe we can tuck the array check in `getKnotIndexOfInvokeCacheArrayAppendixElement`

nbhuiyan replied:
Similar explanation as described in [my other response](https://github.com/eclipse-openj9/openj9/pull/17625#discussion_r1239254436).

mpirvu reviewed (commented):
I am having second thoughts about the implementation. The code seems to be peeking inside objects without making sure it has VM access. Without VM access in hand, the GC can move the objects while the compiler attempts to inspect them. I made some suggestions on how the code needs to be changed.
Moreover, the JITServer should never receive any object references. Thus, the new frontend method should just fatal assert if executed at the server rather than sending a message at the client.

mpirvu commented:
I am surprised though that we cache object references in J9Class (J9Class->callSites[]). Either GC cannot move such objects or somehow it updates all these pointers when objects are moved.

nbhuiyan commented:
@mpirvu I have updated the implementation based on the review comments. I would appreciate another review.

mpirvu commented on runtime/compiler/control/JITClientCompilationThread.cpp:
```diff
@@ -1098,6 +1098,11 @@ handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe, JITServer::Mes
          client->write(response, fe->targetMethodFromMethodHandle(comp, std::get<0>(recv)));
          }
          break;
+      case MessageType::VM_isInvokeCacheEntryAnArray:
```
Is this used?

mpirvu commented on runtime/compiler/net/MessageTypes.cpp:
```diff
@@ -187,6 +187,7 @@ const char *messageNames[] =
    "VM_getVMTargetOffset",
    "VM_getVMIndexOffset",
    "VM_inSnapshotMode",
+   "VM_isInvokeCacheEntryAnArray",
```
Is the new message needed?

mpirvu commented on runtime/compiler/net/MessageTypes.hpp:
```diff
@@ -196,6 +196,7 @@ enum MessageType : uint16_t
    VM_getVMTargetOffset,
    VM_getVMIndexOffset,
    VM_inSnapshotMode,
+   VM_isInvokeCacheEntryAnArray,
```
Is the new message needed?

mpirvu commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -899,6 +899,17 @@ class TR_J9VMBase : public TR_FrontEnd
+    * \param invokeCacheArray the invokeCacheArray address
+    * \return true if entry is an array type; false otherwise
+    */
+   virtual bool isInvokeCacheEntryAnArray(uintptr_t *invokeCacheArray);
```
Maybe using `j9object_t` instead of the generic `uintptr_t` is more suggestive here.

nbhuiyan replied:
I agree, but `uintptr_t *` is used in every function that takes `invokeCacheArray` address, as well as all other VM frontend functions that take an object location.

mpirvu commented on runtime/compiler/env/VMJ9Server.cpp:
```diff
@@ -2228,6 +2228,17 @@ TR_J9ServerVM::targetMethodFromMethodHandle(TR::Compilation* comp, TR::KnownObje
+      JITServer::MessageType::VM_getKnotIndexOfInvokeCacheArrayAppendixElement,
+      invokeCacheArray
+      );
+   return std::get<0>(stream->read<bool>());
```
The response from the client does not appear to be a bool. Am I missing something?

nbhuiyan replied:
This was supposed to be `VM_isInvokeCacheEntryAnArray`, but while squashing and rebasing I might have ordered one of the commits incorrectly. I have fixed this to use the right message, where the response type is bool. This also addresses your other comments about whether `VM_isInvokeCacheEntryAnArray` message is used anywhere.

mpirvu reviewed (approved):
LGTM

mpirvu commented:
jenkins test sanity all jdk20

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk20

mpirvu commented:
All tests have passed.
@0xdaryl if you officially approve this PR I can merge it. Thanks

0xdaryl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6605,6 +6605,14 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
       {
       TR_OpaqueMethodBlock * targetJ9MethodBlock = NULL;
       uintptr_t * invokeCacheArray = (uintptr_t *) callSiteTableEntryAddress(callSiteIndex);
+         // invokedynamic resolution can either result in a valid entry in the corresponding CallSite table slot if successful,
```
Please fix comment indenting here.

0xdaryl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6605,6 +6605,14 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
+         // The exception object would not be an array type object, and therefore it is sufficient to check if
+         // the entry obtained from the CallSite table is an array instance to determine if the resolution was successful.
+      if (!fej9()->isInvokeCacheEntryAnArray(invokeCacheArray))
+            comp->failCompilation<TR::CompilationException>("Invalid CallSite table entry for invokedynamic");
```
Can you add curly braces around this and fix indenting?  The proximity to the following scope makes it appear that scope goes with this "if".

nbhuiyan commented:
@0xdaryl I have addressed your comments. The [latest force push](https://github.com/eclipse-openj9/openj9/compare/b314536921ed4a501ca2b987a0c6c5a76d0feba1..c67008bff023ed4624c76ebce5da272444c1c8d8) addressed indentation issues, removed an empty line, and added curly braces for an if statement. As the tests have passed earlier although not visible due to the updated branch, relaunching them should not be necessary for these changes alone.

0xdaryl commented:
Recent changes were mostly whitespace.  Previous testing was successful.  No need to re-run CI.  Merging.
